File: stuff/pyfemkit/coreClasses/dofClass.py
import numpy as np
from scipy.sparse import csr_matrix
from commonFunctions.lagrangeShapeFunctions import lagrangeShapeFunctions

class dofClass(object):
    def __init__(self):
        self.qN = np.array([],dtype=float)
        self.qN1 = np.array([],dtype=float)
        self.vN = np.array([],dtype=float)
        self.vN1 = np.array([],dtype=float)
        self.listContinuumObjects = []
        self.totalNumberOfDofs = 0
        # self.dirichletDof
        self.numberOfContinuumObjects
        # self.solveDof
        self.callMassMatrixElement

    @property
    def numberOfContinuumObjects(self):
        return len(self.listContinuumObjects)
        
    # @property
    def dirichletDof(self):
        dirichletDofArray = []
        for index1 in range(self.numberOfContinuumObjects):
            continuumObject = self.listContinuumObjects[index1]
            if hasattr(continuumObject,'globalDirichletList'):  
                dirichletDofArray = np.unique(np.concatenate((dirichletDofArray,continuumObject.globalDirichletList()), axis = None))
        return dirichletDofArray
    
    # @property
    def solveDof(self):
        solveDofArray = list(range(self.totalNumberOfDofs))
        if len(self.dirichletDof())!=0:
            for index1 in range(len(self.dirichletDof())):
                solveDofArray.remove(self.dirichletDof()[index1])
            return solveDofArray

    def initialize(self,fieldNameCell):
        for index1 in range(len(fieldNameCell)):
            fieldName = fieldNameCell[index1]
            assert isinstance(fieldName, str),'input need to be of type string'  
            setattr(self,fieldName,np.zeros(self.totalNumberOfDofs))

    def updateGlobalField(self,fieldNameCell):
        for index1 in range(self.numberOfContinuumObjects):
            continuumObject = self.listContinuumObjects[index1]
            for index2 in range(len(fieldNameCell)):
                fieldName = fieldNameCell[index2]
                assert isinstance(fieldName, str),'input need to be of type string'               
                if hasattr(continuumObject,fieldName):
                    getattr(self,fieldName)[continuumObject.globalNodesDof] = getattr(continuumObject,fieldName)
            ## Test
            # self.vN1 = np.array([1,1,1.0,1,1,1])
            # fieldName2 = 'vN1'
            # ind1 = np.array([1,3,5],int)
            # ind2 = np.array([0,1,2],int)
                # if hasattr(self,fieldName2):
                    # getattr(self,fieldName)[ind1] = getattr(self,fieldName2)[ind2]

    def updateTimeDependentField(self,fieldNameCell,time,flag):
        for index1 in range(self.numberOfContinuumObjects):
            continuumObject = self.listContinuumObjects[index1]
            for index2 in range(len(fieldNameCell)):
                fieldName = fieldNameCell[index2]
                assert isinstance(fieldName, str),'input need to be of type string'               
                if hasattr(continuumObject,fieldName):
                    if continuumObject.flagNewton[flag]:
                        continuumObject.time = time
                        getattr(self,fieldName)[continuumObject.globalDirichletList()] = getattr(continuumObject,fieldName)()
                        
    def updateContinuumField(self,fieldNameCell,flag):
        for index1 in range(self.numberOfContinuumObjects):
            continuumObject = self.listContinuumObjects[index1]
            for index2 in range(len(fieldNameCell)):
                fieldName = fieldNameCell[index2]
                assert (isinstance(fieldName, str) and isinstance(flag, str)),'input need to be of type string'
                field = getattr(self,fieldName)
                if (hasattr(continuumObject,fieldName) and flag in continuumObject.flagNewton):
                    if continuumObject.flagNewton[flag]:
                        setattr(continuumObject,fieldName,field[continuumObject.globalNodesDof])

    @property
    def callMassMatrixElement(self):
        # massmassDataFEData = []
        for index1 in range(self.numberOfContinuumObjects):
            continuumObject = self.listContinuumObjects[index1]
            if (hasattr(continuumObject,'massMatrix')):  
                # TODO
                # massData = [massData, continuumObject.massMatrix()]
                massDataFE = continuumObject.massMatrix()
            # assembly
            return csr_matrix((np.concatenate(massDataFE['MeVector']),(np.concatenate(massDataFE['indexMi']), np.concatenate(massDataFE['indexMj'])))) 
    
    def initializeGlobalDofs(self,flag):
        for index1 in range(self.numberOfContinuumObjects):
            continuumObject = self.listContinuumObjects[index1]
            if flag in continuumObject.flagNewton:
                if continuumObject.flagNewton[flag]:
                    numberOfDofPerNode = continuumObject.nodes.shape[1]
                    edof = continuumObject.edof
                    numberOfNodes = continuumObject.nodes.shape[0]
                    numberOfElements = edof.shape[0]
                    globalFullEdof = np.zeros((numberOfElements,edof.shape[1]*numberOfDofPerNode))
                    for j in range(numberOfDofPerNode):
                        globalFullEdof[:,j::numberOfDofPerNode] = self.totalNumberOfDofs + (edof+1)*numberOfDofPerNode - (numberOfDofPerNode-(j+1))
                    globalFullEdof = np.array(globalFullEdof - 1,dtype=int)
                    # globalFullEdof(:,j:numberOfDofPerNode:end) = obj.totalNumberOfDofs + edof*numberOfDofPerNode-(numberOfDofPerNode-j);
                    globalNodesDof = np.zeros((numberOfNodes,numberOfDofPerNode))
                    globalNodesDof[:,0:numberOfDofPerNode+1] = self.totalNumberOfDofs + np.kron(np.ones(numberOfNodes).reshape(-1,1),range(1,numberOfDofPerNode+1)) + np.kron(np.arange(0,numberOfDofPerNode*(numberOfNodes),numberOfDofPerNode).reshape([-1,1]),np.ones(numberOfDofPerNode))
                    #
                    globalNodesDof = np.array(globalNodesDof - 1,dtype=int)
                    #
                    continuumObject.globalFullEdof = globalFullEdof
                    continuumObject.globalNodesDof = globalNodesDof
                    self.totalNumberOfDofs = self.totalNumberOfDofs + numberOfNodes*numberOfDofPerNode

    def initializeQV(self,flag):
        for index1 in range(self.numberOfContinuumObjects):
            continuumObject = self.listContinuumObjects[index1]
            if flag in continuumObject.flagNewton:
                if continuumObject.flagNewton[flag]:
                    if continuumObject.qN.size == 0:
                        continuumObject.qN = continuumObject.qR
                    if continuumObject.qN1.size == 0:
                        continuumObject.qN1 = continuumObject.qR
                    if continuumObject.vN.size == 0:
                        continuumObject.vN = np.zeros(continuumObject.nodes.shape)
                        
    def initializeShapeFunctions(self,flag):
        for index1 in range(self.numberOfContinuumObjects):
            continuumObject = self.listContinuumObjects[index1]
            if flag in continuumObject.flagNewton:
                if continuumObject.flagNewton[flag]:
                    continuumObject.shapeFunctions = lagrangeShapeFunctions(continuumObject.edof.shape[1], continuumObject.numberOfGausspoints, continuumObject.dimension)

File: stuff/tag/2021_03_16/coreClasses/dofClass.m
classdef dofClass < handle
    properties
        qN
        qN1
        vN
        vN1
        listContinuumObjects
        totalNumberOfDofs = 0;
        dirichletDof
    end
    properties (Dependent = true)
        numberOfContinuumObjects
        callMassMatrixElement
        solveDof
    end
    methods
        % constructor
        function obj = dofClass
        end
    end
    methods        
        function dirichletDof = get.dirichletDof(obj)
            dirichletDof = [];
            for index1 = 1:obj.numberOfContinuumObjects
                continuumObject = obj.listContinuumObjects{index1};
                if ~isempty(continuumObject)
                    if isprop(continuumObject,'globalDirichletList')
                        dirichletDof = unique([dirichletDof; continuumObject.globalDirichletList]);
                    end
                end
            end
        end
        function solveDof = get.solveDof(obj)
            solveDof = 1:obj.totalNumberOfDofs;
            if ~isempty(obj.dirichletDof)
                solveDof(obj.dirichletDof(:,1)) = [];
            end
        end    
        function out = get.numberOfContinuumObjects(obj)
            out = numel(obj.listContinuumObjects);
        end
        function initialize(obj,fieldNameCell)
            for index1 = 1:numel(fieldNameCell)
                fieldName = fieldNameCell{index1};
                assert(ischar(fieldName));
                obj.(fieldName) = zeros(obj.totalNumberOfDofs,1);
            end
        end
        function updateGlobalField(obj,fieldNameCell)
            for index1 = 1:obj.numberOfContinuumObjects
                continuumObject = obj.listContinuumObjects{index1};
                for index2 = 1:numel(fieldNameCell)
                    fieldName = fieldNameCell{index2};
                    assert(ischar(fieldName));
                    if isprop(continuumObject,fieldName)
                        obj.(fieldName)(continuumObject.globalNodesDof) = continuumObject.(fieldName);
                    end
                end
            end
        end
        function updateTimeDependentField(obj,fieldNameCell,time,flagUpdate) 
            for index1 = 1:obj.numberOfContinuumObjects
                continuumObject = obj.listContinuumObjects{index1};
                fieldName = fieldNameCell{1};
                assert(ischar(fieldName));
                if isprop(continuumObject,fieldName)
                    if continuumObject.flagNewton.(flagUpdate)
                        continuumObject.time = time;
                        obj.(fieldName)(continuumObject.globalDirichletList) = continuumObject.(fieldName);
                    end
                end
            end
        end        
        function updateContinuumField(obj,fieldNameCell,flagUpdate)
            for index1 = 1:obj.numberOfContinuumObjects
                continuumObject = obj.listContinuumObjects{index1};
                for index2 = 1:numel(fieldNameCell)
                    fieldName = fieldNameCell{index2};
                    field = obj.(fieldName);
                    assert(ischar(fieldName) & ischar(flagUpdate));
                    if isprop(continuumObject,fieldName) && isfield(continuumObject.flagNewton,flagUpdate)
                        if continuumObject.flagNewton.(flagUpdate)
                            continuumObject.(fieldName) = field(continuumObject.globalNodesDof);
                        end
                    end
                end
            end
        end
        function out = get.callMassMatrixElement(obj)
            massData = [];
            for index1 = 1:obj.numberOfContinuumObjects
                continuumObject = obj.listContinuumObjects{index1};
                if isprop(continuumObject,'massMatrix')
                    massData = [massData, continuumObject.massMatrix];
                end
            end
            % assembly
            out = sparse(vertcat(massData(:).indexMi),vertcat(massData(:).indexMj),vertcat(massData(:).MeVector),obj.totalNumberOfDofs,obj.totalNumberOfDofs);
        end
        function initializeGlobalDofs(obj,flag)
            for index1 = 1:obj.numberOfContinuumObjects
                continuumObject = obj.listContinuumObjects{index1};
                if isfield(continuumObject.flagNewton,flag)
                    if continuumObject.flagNewton.(flag)
                        numberOfDofPerNode = size(continuumObject.nodes,2);
                        edof = continuumObject.edof;
                        numberOfNodes = size(continuumObject.nodes,1);
                        numberOfElements = size(edof,1);
                        globalFullEdof = zeros(numberOfElements,size(edof,2)*numberOfDofPerNode);
                        for j = 1:numberOfDofPerNode
                            globalFullEdof(:,j:numberOfDofPerNode:end) = obj.totalNumberOfDofs + edof*numberOfDofPerNode-(numberOfDofPerNode-j);
                        end
                        globalNodesDof = zeros(numberOfNodes,numberOfDofPerNode);
                        globalNodesDof(:,1:numberOfDofPerNode) = obj.totalNumberOfDofs + kron(ones(numberOfNodes,1),1:numberOfDofPerNode)+kron((0:numberOfDofPerNode:numberOfDofPerNode*numberOfNodes-1)',ones(1,numberOfDofPerNode));
                        continuumObject.globalFullEdof = globalFullEdof;
                        continuumObject.globalNodesDof = globalNodesDof;
                        obj.totalNumberOfDofs = obj.totalNumberOfDofs + numberOfNodes*numberOfDofPerNode;
                    end
                end
            end
        end
        function initializeQV(obj,flag)
            for index1 = 1:obj.numberOfContinuumObjects
                continuumObject = obj.listContinuumObjects{index1};
                if isfield(continuumObject.flagNewton,flag)
                    if continuumObject.flagNewton.(flag)
                        if isempty(continuumObject.qN)
                            continuumObject.qN = continuumObject.qR;
                        end
                        if isempty(continuumObject.qN1)
                            continuumObject.qN1 = continuumObject.qR;
                        end
                        if isempty(continuumObject.vN)
                            continuumObject.vN = zeros(size(continuumObject.nodes,1), size(continuumObject.nodes,2));
                        end
                    end
                end
            end
        end
        function initializeShapeFunctions(obj,flag)
            for index1 = 1:obj.numberOfContinuumObjects
                continuumObject = obj.listContinuumObjects{index1};
                if isfield(continuumObject.flagNewton,flag)
                    if continuumObject.flagNewton.(flag)
                        continuumObject.shapeFunctions = lagrangeShapeFunctions(size(continuumObject.edof,2), continuumObject.numberOfGausspoints, continuumObject.dimension);
                    end
                end
            end
        end
    end
end
